File: web-core/src/main/java/cn/labzen/web/response/format/AbnormalStatusResponseFormatter.java
package cn.labzen.web.response.format;

import cn.labzen.web.api.definition.HttpStatusExt;
import cn.labzen.web.api.response.out.Response;
import com.google.common.primitives.Ints;
import jakarta.servlet.RequestDispatcher;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;

import java.util.Map;
import java.util.Optional;

/**
 * 异常状态码格式化器
 * <p>
 * 处理 Spring MVC 产生的异常状态响应（如 404、500 等错误页面）。
 * 将这些异常响应转换为标准的 JSON 响应结构。
 */
public class AbnormalStatusResponseFormatter implements ResponseFormatter {

  private static final int UNEXPECTED_ERROR = HttpStatusExt.UNEXPECTED_ERROR.code();

  /**
   * 支持 Map 类型且包含错误状态码属性
   */
  @Override
  public boolean support(Class<?> clazz, HttpServletRequest request) {
    return Map.class.isAssignableFrom(clazz) && request.getAttribute(RequestDispatcher.ERROR_STATUS_CODE) != null;
  }

  /**
   * 提取错误状态码和消息，转换为标准响应
   */
  @Override
  public Object format(Object result, HttpServletRequest request, HttpServletResponse response) {
    @SuppressWarnings("unchecked") 
    Map<String, Object> data = (Map<String, Object>) result;
    String status = data.getOrDefault("status", "").toString();
    Integer code = Optional.ofNullable(Ints.tryParse(status)).orElse(UNEXPECTED_ERROR);
    String message = data.getOrDefault("error", "").toString();
    return new Response(code, message, null, null);
  }
}

File: src/main/kotlin/cn/labzen/web/response/format/AbnormalStatusResponseFormatter.kt
package cn.labzen.web.response.format

import cn.labzen.web.response.bean.Response
import javax.servlet.RequestDispatcher
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

/**
 * 格式化不正常的 Http Status 结果，如404
 */
class AbnormalStatusResponseFormatter : ResponseFormatter {

  override fun support(clazz: Class<*>, request: HttpServletRequest): Boolean =
    clazz == Map::class.java && request.getAttribute(RequestDispatcher.ERROR_STATUS_CODE) != null

  override fun format(result: Any?, request: HttpServletRequest, response: HttpServletResponse): Any {
    val data = result as Map<*, *>
    val code = data["status"].toString().toInt()
    val message = data["error"].toString()
    return Response(code, message)
  }
}